### PR TITLE
[SU-25] Refactor EntityEditor

### PR DESCRIPTION
**Pull request by nawatts**
The data table allows editing the value of an attribute using the EntityEditor component. This splits up EntityEditor and extracts the form for entering the attribute's value/type so that they can be reused for adding new rows (in a future PR).

### CODE REVIEW

**Comment by cahrens** [comment]
@nawatts is this PR a refactor in support of SU-25, or does it also add the ability to add rows/columns to the data table? I'm running the deployment but am not sure I know what I'm looking or.

**Comment by nawatts** [comment]
Sorry, should have clarified. This is in support of SU-25 and adding new rows would be added in a future PR. This PR should not make any visible changes.

File: src/components/data/_tests/data-utils.test.js [hunk 1]
```diff
@@ -61,3 +61,79 @@ describe('concatenateAttributeNames', () => {
...
+    expect(getAttributeType('value')).toEqual({ type: 'string', isList: false })
+    expect(getAttributeType(3)).toEqual({ type: 'number', isList: false })
+    expect(getAttributeType(false)).toEqual({ type: 'boolean', isList: false })
+    expect(getAttributeType({
+      entityType: 'thing',
+      entityName: 'thing_one'
+    })).toEqual({ type: 'reference', isList: false })
+
+    expect(getAttributeType({ items: ['a', 'b', 'c'] })).toEqual({ type: 'string', isList: true })
+    expect(getAttributeType({ items: [1, 2, 3] })).toEqual({ type: 'number', isList: true })
```
Comment (cahrens) [inline src/components/data/_tests/data-utils.test.js]: What does it do with mixed types (`['a', 2, false]`, for example)?

File: src/components/data/_tests/data-utils.test.js [hunk 2]
```diff
@@ -61,3 +61,79 @@ describe('concatenateAttributeNames', () => {
...
+      items: [
+        { entityType: 'thing', entityName: 'thing_one' },
+        { entityType: 'thing', entityName: 'thing_two' }
+      ],
+      itemsType: 'EntityReference'
+    })).toEqual({ type: 'reference', isList: true })
+  })
+
+  it('returns string for null values', () => {
+    expect(getAttributeType(null)).toEqual({ type: 'string', isList: false })
```
Comment (cahrens) [inline src/components/data/_tests/data-utils.test.js:88]: These tests are great! Some of these conversions are not at all obvious, and it is great to have documentation of what they do.

File: src/components/data/_tests/data-utils.test.js (hunk 1 above)
Comment (nawatts) [reply src/components/data/_tests/data-utils.test.js]: I _think_ that a list is not allowed to contain items of mixed types.

However, I'm not sure if it's possible to create a generic JSON value that looks like a list. Generic JSON values currently break this component ([AJ-342](https://broadworkbench.atlassian.net/browse/AJ-342)).

Comment (nawatts) [reply src/components/data/_tests/data-utils.test.js]: Rawls does reject creating a list with mixed types. `batchUpsert` with:
```
[
  {
    "name": "thing_one",
    "entityType": "things",
    "operations": [
      {
        "op": "CreateAttributeValueList",
        "attributeName": "new_attr"
      },
      {
        "op": "AddListMember",
        "attributeListName": "new_attr",
        "newMember": "a"
      },
      {
        "op": "AddListMember",
        "attributeListName": "new_attr",
        "newMember": 2
      },
      {
        "op": "AddListMember",
        "attributeListName": "new_attr",
        "newMember": false
      }
    ]
  }
]
```
returns an error:
> inconsistent attributes for list: attribute lists must consist of a single data type. For attribute 'new_attr', found types: [AttributeString, AttributeNumber, AttributeBoolean]. Sample values for these types: [AttributeString(a), AttributeNumber(2), AttributeBoolean(false)]

There is still the issue of creating a raw JSON attribute that looks like a list, but I think maybe we can defer that to AJ-342 since this is the same as the current behavior of EntityEditor: https://github.com/DataBiosphere/terra-ui/blob/9053e0a383676f8ac183d278204c3e2ae35066be/src/components/data/data-utils.js#L478-L485

File: src/components/data/data-utils.js [hunk 3]
```diff
@@ -472,68 +472,217 @@ export const EntityRenamer = ({ entityType, entityName, workspaceId: { namespace
...
+        value,
+        onChange: v => onChange(_.trim(v)),
+        ...otherProps
+      })
+    }],
+    ['reference', () => {
+      const { value, onChange, ...otherProps } = props
+      return h(TextInput, {
+        autoFocus: true,
+        placeholder: 'Enter a value',
```
Comment (cahrens) [inline src/components/data/data-utils.js]: This used to be `Enter a ${linkedEntityType}_id`. Was the change intentional?

Comment (nawatts) [reply src/components/data/data-utils.js]: Oops, that was not intentional.

Comment (nawatts) [reply src/components/data/data-utils.js]: Fixed in [5342971](https://github.com/DataBiosphere/terra-ui/pull/2895/commits/5342971e15085e1c826f60e9c87852c1c551f3df).

File: src/components/data/data-utils.js [hunk 4]
```diff
@@ -472,68 +472,217 @@ export const EntityRenamer = ({ entityType, entityName, workspaceId: { namespace
...
+      return _.isNaN(numberVal) ? 0 : numberVal
+    }],
+    [Utils.DEFAULT, () => Utils.convertValue(newType)]
   )
+  const convertFn = type === 'reference' ? _.flow(_.get('entityName'), baseConvertFn) : baseConvertFn
+
+  if (isList) {
+    return _.flow(
+      _.update('items', _.map(convertFn)),
+      newType === 'reference' ? _.set('itemsType', 'EntityReference') : _.omit('itemsType')
```
Comment (cahrens) [inline src/components/data/data-utils.js]: I don't understand why `itemsType` would be present (the omit part of this statement). Can you add a comment about what this code is doing?

Comment (nawatts) [reply src/components/data/data-utils.js]: Lists of entity references look like:
```
{
   items: [
     { entityType: 'thing', entityName: 'thing_1' },
     { entityType: 'thing', entityName: 'thing_1' }
  ],
  itemsType: 'EntityReference'
}
```

This was dropping that `itemsType` field when converting from a list of references to a list of some other type. However, I missed that lists of other types also have an `itemsType` field, but it's set to `AttributeValue` instead of `EntityReference`.

[e8870fd](https://github.com/DataBiosphere/terra-ui/pull/2895/commits/e8870fdb45a008185b03a88c1cb6db2bfd6ad678) adds `itemsType` for non-reference types and makes this conversion clearer.

File: src/components/data/data-utils.js [hunk 5]
```diff
@@ -472,68 +472,215 @@ export const EntityRenamer = ({ entityType, entityName, workspaceId: { namespace
...
+  )
+}
+
+const AttributeInput = ({ value: attributeValue, onChange, entityTypes = [] }) => {
+  const { type: attributeType, isList } = getAttributeType(attributeValue)
+
+  const renderInput = renderInputForAttributeType(attributeType)
+
+  const defaultReferenceEntityType = Utils.cond(
+    [attributeType === 'reference' && isList, () => !_.isEmpty(attributeValue.items) ? attributeValue.items[0].entityType : entityTypes[0]],
```
Comment (cahrens) [inline src/components/data/data-utils.js:572]: It looks like this used to return undefined. Was this change intentional?

```
const [linkedEntityType, setLinkedEntityType] = useState(() => Utils.cond(
    [initialIsReference && initialIsList, () => attributeValue.items[0] ? attributeValue.items[0].entityType : undefined],
    [initialIsReference, () => attributeValue.entityType],
    () => entityTypes[0]
  ))

Comment (nawatts) [reply src/components/data/data-utils.js:572]: This one was intentional. The new `defaultReferenceEntityType` variable is used slightly differently from the old `linkedEntityType` variable. While an empty entity reference list does not specify an entity type, we still need some entity type to use for added list items. So if we can't determine an entity type from the attribute, we pick one from the available entity types in the workspace.

**Review by cahrens — APPROVED** [review]
LGTM, though I have not tested the code. @MatthewBemis please do some bashing.

**Review by MatthewBemis — APPROVED** [review]
This seems good to me. I played around with it, no issues.
